### DISCUSSION TITLE
the noble/ed25519 library returns "hex invalid" error when signing private keys

### DISCUSSION THREAD

**Discussion by Aixoxa** [paulmillr/noble-ed25519#101] [General]
Am making use of the orderly api for authentication which uses the noble/ed25519 library for signing private keys but everytime i run the code the noble/ed25519 returns this error "hex invalid" I've tried with the node web crypto for generating ed25519  keys still returns the same error, please is there a fix for this?

Here's the error:

![Screenshot 2024-05-20 085222](https://github.com/paulmillr/noble-ed25519/assets/53108906/b3aae564-dace-45de-949a-fc5f35ae7de3)

Here's the code below:

import { getPublicKeyAsync, signAsync } from '@noble/ed25519';
import { encodeBase58 } from 'ethers';

export async function signAndSendRequest(
  orderlyAccountId: string,
  privateKey: Uint8Array | string,
  input: URL | string,
  init?: RequestInit | undefined
): Promise<Response> {
  const timestamp = Date.now();
  const encoder = new TextEncoder();

  const url = new URL(input);
  let message = `${String(timestamp)}${init?.method ?? 'GET'}${url.pathname}${url.search}`;
  if (init?.body) {
    message += init.body;
  }
  const orderlySignature = await signAsync(encoder.encode(message), privateKey);

  return fetch(input, {
    headers: {
      'Content-Type':
        init?.method !== 'GET' && init?.method !== 'DELETE'
          ? 'application/json'
          : 'application/x-www-form-urlencoded',
      'orderly-timestamp': String(timestamp),
      'orderly-account-id': orderlyAccountId,
      'orderly-key': `ed25519:${encodeBase58(await getPublicKeyAsync(privateKey))}`,
      'orderly-signature': Buffer.from(orderlySignature).toString('base64url'),
      ...(init?.headers ?? {})
    },
    ...(init ?? {})
  });
}

**Comment by paulmillr** [comment]
The error means the hex string you've passed in privateKey is invalid. It probably has some weird non-hex characters (hex is 0-9 a-f), or is not padded (abcdef vs 0abcdef).

**Reply by Aixoxa** [reply]
I don't understand 
the orderly api key generated from the logx was this ed25519:B4yEc8xMjoqdUiKT5ZT1b2cUy5cggZrbMPvCrqyjkQhf

**Reply by paulmillr** [reply]
Read up what's "hex string", and what's "base64-encoded strong".

**Reply by Aixoxa** [reply]
I know what they are but the orderly API key generated for me from logx is this ed25519:7tEoJo5hMBKBrQsqmc8yw1xNfoQCFQBVwQT1eFafRNRf so please how do i fix this with the @noble/ed25519 library

**Reply by Aixoxa** [reply]
Thank you so much Paul I finally fixed it.
